### PR TITLE
3617: Suppliers - Edit Native Desktop - Supported Operating Systems

### PR DESCRIPTION
**Pull request by assoomro**
- PUT endpoint together with corresponding tests

### CODE REVIEW

**Comment by sam-roberts-bjss** [comment]
The conflicts will need resolving first. Also could you add the Story ID to the commit message?

**Comment by assoomro** [comment]
> The conflicts will need resolving first. Also could you add the Story ID to the commit message?

Conflicts resolved, does that rename do? :]

File: src/NHSD.BuyingCatalogue.Solutions.Application/Commands/NativeDesktop/UpdateSolutionNativeDesktopOperatingSystems/UpdateSolutionNativeDesktopOperatingSystemsCommand.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    public sealed class UpdateSolutionNativeDesktopOperatingSystemsCommand : IRequest<ISimpleResult>
+    {
+        public string SolutionId { get; }
+
+        public string NativeDesktopOperatingSystemsDescription { get; }
+
+        public UpdateSolutionNativeDesktopOperatingSystemsCommand(string solutionId, string nativeDesktopOperatingSystemsDescription)
+        {
+            SolutionId = solutionId.ThrowIfNull();
+            NativeDesktopOperatingSystemsDescription = nativeDesktopOperatingSystemsDescription?.Trim(); ;
```
Comment (CallumBradburyBJSS) [inline src/NHSD.BuyingCatalogue.Solutions.Application/Commands/NativeDesktop/UpdateSolutionNativeDesktopOperatingSystems/UpdateSolutionNativeDesktopOperatingSystemsCommand.cs]: there is an extra semicolon at the end of this line

File: src/NHSD.BuyingCatalogue.Solutions.Contracts/IClientApplication.cs [hunk 2]
```diff
@@ -20,5 +20,7 @@ public interface IClientApplication
         IMobileMemoryAndStorage MobileMemoryAndStorage { get; }
         IMobileThirdParty MobileThirdParty { get; }
         string NativeMobileAdditionalInformation { get; }
+
+        string NativeDesktopOperatingSystemsDescription { get; }
```
Comment (CallumBradburyBJSS) [inline src/NHSD.BuyingCatalogue.Solutions.Contracts/IClientApplication.cs]: consistency is key, remove the newline

File: tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/BrowserBased/BrowserAdditionalInformation/Supplier Edit BrowserAdditionalInformation.3601.feature [hunk 3]
```diff
@@ -24,7 +24,7 @@ Scenario: 1. Browser Additional Information is updated
...
+        | Sln1     | An full online medicine system | Online medicine 1 | { "ClientApplicationTypes": ["browser-based"],"BrowsersSupported" : [ "IE8", "Opera" ], "MobileResponsive": false, "Plugins": null, "MinimumConnectionSpeed": null, "MinimumDesktopResolution": null, "HardwareRequirements": "New Hardware", "NativeMobileHardwareRequirements": null, "AdditionalInformation": "New Additional Info", "MobileFirstDesign": null, "NativeMobileFirstDesign": null, "MobileOperatingSystems": null, "MobileConnectionDetails": null, "MobileMemoryAndStorage": null, "MobileThirdParty": null, "NativeMobileAdditionalInformation": null, "NativeDesktopOperatingSystemsDescription": null } |
```
Comment (CallumBradburyBJSS) [inline tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/BrowserBased/BrowserAdditionalInformation/Supplier Edit BrowserAdditionalInformation.3601.feature]: fix the |

File: tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/BrowserBased/BrowserHardwareRequirements/Supplier Edit BrowserHardwareRequirements.3600.feature [hunk 4]
```diff
@@ -24,7 +24,7 @@ Scenario: 1. Browser Hardware Requirements is updated
...
+        | Sln1     | An full online medicine system | Online medicine 1 | { "ClientApplicationTypes": ["browser-based"],"BrowsersSupported" : [ "IE8", "Opera" ], "MobileResponsive": false, "Plugins": null, "MinimumConnectionSpeed": null, "MinimumDesktopResolution": null, "HardwareRequirements": "New Hardware", "NativeMobileHardwareRequirements": null, "AdditionalInformation": null, "MobileFirstDesign": null, "NativeMobileFirstDesign": null, "MobileOperatingSystems": null, "MobileConnectionDetails": null, "MobileMemoryAndStorage": null, "MobileThirdParty": null, "NativeMobileAdditionalInformation": null, "NativeDesktopOperatingSystemsDescription": null } |
```
Comment (CallumBradburyBJSS) [inline tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/BrowserBased/BrowserHardwareRequirements/Supplier Edit BrowserHardwareRequirements.3600.feature]: fix the |

File: tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/BrowserBased/BrowserMobileFirst/Supplier Edit BrowserMobileFirst.3602.feature [hunk 5]
```diff
@@ -24,7 +24,7 @@ Scenario: 1. Browser Mobile First is updated
...
+        | Sln1     | An full online medicine system | Online medicine 1 | { "ClientApplicationTypes": ["browser-based"],"BrowsersSupported" : [ "IE8", "Opera" ], "MobileResponsive": false, "Plugins": null, "MinimumConnectionSpeed": null, "MinimumDesktopResolution": null, "HardwareRequirements": "New Hardware", "NativeMobileHardwareRequirements": null, "AdditionalInformation": "Some Info", "MobileFirstDesign": true, "NativeMobileFirstDesign": null, "MobileOperatingSystems": null, "MobileConnectionDetails": null, "MobileMemoryAndStorage": null, "MobileThirdParty": null, "NativeMobileAdditionalInformation": null, "NativeDesktopOperatingSystemsDescription": null } |
```
Comment (CallumBradburyBJSS) [inline tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/BrowserBased/BrowserMobileFirst/Supplier Edit BrowserMobileFirst.3602.feature]: fix the |

File: tests/NHSD.BuyingCatalogue.Solutions.API.UnitTests/NativeDesktop/NativeDesktopOperatingSystemsControllerTests.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        private const string SolutionId = "Sln1";
+
+        [SetUp]
+        public void Setup()
+        {
+            _mockMediator = new Mock<IMediator>();
+            _desktopOperatingSystemsController = new NativeDesktopOperatingSystemsController(_mockMediator.Object);
+        }
+        [TestCase("d")]
+        [TestCase("all sorts of characters, up to 1000, but no more.")]
```
Comment (CallumBradburyBJSS) [inline tests/NHSD.BuyingCatalogue.Solutions.API.UnitTests/NativeDesktop/NativeDesktopOperatingSystemsControllerTests.cs]: these test cases aren't testing anything.

The way the controller tests work, it doesn't matter what gets passed down into mediatr, as the controller isn't doing any validation itself.

You can get away with a single test for each case, as you're mocking out the ISimpleResult. If it's mocked to Valid it should succeed, if it's mocked to Invalid it should fail. The nuances of what causes Valid/Invalid aren't these tests concern.

File: src/NHSD.BuyingCatalogue.Solutions.Application/Commands/NativeDesktop/UpdateSolutionNativeDesktopOperatingSystems/UpdateSolutionNativeDesktopOperatingSystemsCommand.cs (hunk 1 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Solutions.Application/Commands/NativeDesktop/UpdateSolutionNativeDesktopOperatingSystems/UpdateSolutionNativeDesktopOperatingSystemsCommand.cs]: ✔️

File: src/NHSD.BuyingCatalogue.Solutions.Contracts/IClientApplication.cs (hunk 2 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Solutions.Contracts/IClientApplication.cs]: ✔️

File: tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/BrowserBased/BrowserAdditionalInformation/Supplier Edit BrowserAdditionalInformation.3601.feature (hunk 3 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/BrowserBased/BrowserAdditionalInformation/Supplier Edit BrowserAdditionalInformation.3601.feature]: ✔️

File: tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/BrowserBased/BrowserHardwareRequirements/Supplier Edit BrowserHardwareRequirements.3600.feature (hunk 4 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/BrowserBased/BrowserHardwareRequirements/Supplier Edit BrowserHardwareRequirements.3600.feature]: ✔️

File: tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/BrowserBased/BrowserMobileFirst/Supplier Edit BrowserMobileFirst.3602.feature (hunk 5 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.API.IntegrationTests/Features/BrowserBased/BrowserMobileFirst/Supplier Edit BrowserMobileFirst.3602.feature]: ✔️

File: tests/NHSD.BuyingCatalogue.Solutions.API.UnitTests/NativeDesktop/NativeDesktopOperatingSystemsControllerTests.cs (hunk 6 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Solutions.API.UnitTests/NativeDesktop/NativeDesktopOperatingSystemsControllerTests.cs]: Good point

**Pull request merged**